docs: add introduction.md (#354)
namecheap/ilc: docs/routing/introduction.md

# Introduction

Most JavaScript frameworks come with a dedicated routing solution
like `angular/router` or `vue-router`. They make it possible to
navigate through different pages of an application without having to
do a full page refresh on every link click.
Since we no longer have one monolithic application that handles all routes, instead we have several independent applications, we need to solve an important issue, namely routing.

The following describes how this is handled in the ILC, but first let's clear up the terminology:

>  - Hard transitions - Transition in which the application on the page is replaced by another one.

Reviewer: emphasize the fact that one is when routing is handled via ILC, and another is when routing handled inside the fragment

>  - Hard transitions - A transition handled by ILC routing, with such a transition, the application on the page changes to another one.

Reviewer 2: In Manning book "Hard transitions" are the ones that cause full page reload. And I think it's counterintuitive to change the meaning of the statement here. As you still need somehow to call transitions with full page reload.

Also you've missed the whole explanation of the App Shell concept (which is basically what ILC does). As well as at least basic explanation of how ILC actually turns hard transitions into "soft ones" by intercepting one events and re-emitting the others....

Another point: there is a nice visualization of Two-level routing (Figure 7.7) in the book. Would be nice to also have it here with all related theory behind it.

 - Soft transitions - A transition handled by own routing of some application at the page, with such a transition, only the content inside the application changes.

 ![Introdaction demo](../assets/routes/introduction-demo.png)

> #

Reviewer 2: What is this?

Reviewer 2: Looks like unnecessary markup

Now, let's talk a little about the theory, after which we will move on to practical examples:

>  In ILC, we can use one HTML template for all of our applications. With this approach, SSR occurs once when the page is first loaded, then all navigation occurs through CSR. So let's look at navigation in more detail:

Reviewer: "we can use one HTML template" - no, we use always one template, and we don't have the capability to change it. Please correct this sentence.

 For clarity, here is the routing table:
 ![Routing table](../assets/routes/route.png)

> Let's go to the `/people/` route, it will be a `hard transitions` because we render a new application on the `/people/*` route.

Reviewer: in your example it's not a hard transition, it's just SSR. it can confuse people.
you should say something like  "for example, we are at /plants page and want to open '/people' page. in this case we have hard transition". 
correct and rephrase this sentense.

# Introduction

## Problem statement

Most JavaScript frameworks come with a dedicated routing solution like `angular/router` or `vue-router`.

> They allow you to navigate through pages of an application without a full page refresh on every click.

Reviewer 3: Move this under the introduction

> They allow you to navigate through pages of an application without a full page refresh on every click.

Reviewer 3: Move this under the introduction

When you have a monolithic application, it handles all the routes on its own. When there are two or more applications, they also handle all their routes independently. Since, by default, one independent application know nothing about routes and pages of other independent application. This gives you a problem to solve - a routing issue.

> 

Reviewer 3: ```suggestion

## Routing basics

```

!["2-tiered routing" approach](../assets/2_tiered_routing.png)

In the example above, the user opened the `/news/latest` page URL. ILC checks the first part of the URL (`/news/`) to determine the assosiated application. It correlates to the `/news/*` route configured in ILC. This route contains information about applications that should be loaded on the page and props these applications need to receive. When the application is loaded and mounted to its container DOM node, it also receives `basePath` property that should be used by application's router. The application's router processes the complete URL to find the correct page inside its SPA.

You can use native tools (for example `<Link>` in `React router`) to navigate between pages within the application, and `global link` - a link (`<a>` tag) to navigate between applications.

In ILC, as mentioned before, the transition between applications occurs via the `<a>` tags. To do this, ILC keeps track of all `<a>` tags on the page and handles clicks on them, provided that:

> 1. Tag contains a non-empty `href`.

Reviewer 3: Indent a list with a blank line

1. `event.PreventDefault` does not equal `false`.
1. `target`does not equal `_self`.
1. This is not a special url (`mailto`, `tel`, etc).

If one of the above points is not met, ILC ignores the processing of the clicks on the link.

### Conclusion

> ILC acts as a wrapper for other applications making all the transitions soft. Furthermore, it uses two-level routing so that teams can configure routing inside their application as they need to, whereas, in ILC, you only need to specify the path to the application.

Reviewer 3: Indent a heading with a blank line

# Introduction

Most JavaScript frameworks come with a dedicated routing solution
like `angular/router` or `vue-router`. They make it possible to
navigate through different pages of an application without having to
do a full page refresh on every link click.
Since we no longer have one monolithic application that handles all routes, instead we have several independent applications, we need to solve an important issue, namely routing.

The following describes how this is handled in the ILC, but first let's dive a little into the general theory of the micro frontend routing and clear up the terminology:

>  - Hard navigation describes a page transition where the browser

Reviewer 2: ```suggestion
 - **Hard navigation** - describes a page transition where the browser
```

 loads the complete HTML for the next page from the server.

>  - Soft navigation refers to a page transition that’s entirely clientside rendered, typically by using a client-side router. In this

Reviewer 2: ```suggestion
 - **Soft navigation** - refers to a page transition that’s entirely clientside rendered, typically by using a client-side router. In this
```

 scenario the client fetches its data via an API from the server.

>  So, we have several approaches to implement navigation, the image below shows two of them:

Reviewer 2: ```suggestion
 Generally, we have several approaches to implement navigation, the image below shows two most popular of them:
```

  1.  Page transitions happen via plain links, which result in a full refresh of the page. Nothing special is needed - Team A must know how to link to the pages of Team B and vice versa.
  2. All transitions inside team boundaries are soft. Hard navigation happens when the user crosses team boundaries. From an architectural perspective, it’s identical to the first approach. The fact that a team uses a SPA for its pages is an implementation detail. As long as it responds correctly to URLs, the other team doesn’t have to care.

 ![Introdaction demo](../assets/routes/introduction-demo.png)

> At ILC, we use a third approach called Unified SPA*.

Reviewer 2: ```suggestion
In ILC, we use third approach called **Unified SPA**.
```

> ILC will appear in it as an application shell, it's job is to map URLs to the correct team.

Reviewer 2: This sentence is meaningless...

>  * The Unified SPA (Single Page Application) approach introduces a central application container. It handles page transitions between the teams. Here all navigations are soft.

Reviewer 2: Useless bulletpoint... Why are you making lists which contain 1 item only?

 ![Introdaction demo](../assets/routes/introduction-demo2.png)

Now more detail about ILC:

>  In ILC, we can use one HTML template for all of our applications. With this approach, SSR occurs once when the page is first loaded, then all navigation occurs through CSR. In addition to the fact that all navigation inside the ILC is soft, it also uses "2-tiered routing".

Reviewer 2: SSR is an optional thing in ILC. Also it has nothing to do with routing

# Introduction

Most JavaScript frameworks come with a dedicated routing solution
like `angular/router` or `vue-router`. They make it possible to
navigate through different pages of an application without having to
do a full page refresh on every link click.

> Since we no longer have one monolithic application that handles all routes, instead we have several independent applications, we need to solve an important issue, namely routing.

Reviewer 3: Update:

## Problem statement
When you have a monolithic application, it handles all the routes on its own. When there are two or more applications, they also handle all their routes independently. Since, by default, one independent application know nothing about routes and pages of other independent application. This gives you a problem to solve - a routing issue.

 scenario the client fetches its data via an API from the server.

 Generally, we have several approaches to implement navigation, the image below shows two most popular of them:

  1. Page transitions happen via plain links, which result in a full refresh of the page. Nothing special is needed - Team A must know how to link to the pages of Team B and vice versa.
  2. All transitions inside team boundaries are soft. Hard navigation happens when the user crosses team boundaries. From an architectural perspective, it’s identical to the first approach. The fact that a team uses a SPA for its pages is an implementation detail. As long as it responds correctly to URLs, the other team doesn’t have to care.

 ![Introdaction demo](../assets/routes/introduction-demo.png)

In ILC, we use third approach called **Unified SPA**. The Unified SPA (Single Page Application) approach introduces a central application container (it's ILC). It handles page transitions between the teams. Here all navigations are soft.

 ![Introdaction demo](../assets/routes/introduction-demo2.png)

Now more detail about ILC:

>  In ILC, we can use one HTML template for all of our applications. With this approach, initial page loading occurs only once when the page is first time loaded, then all navigation occurs through CSR. In addition to the fact that all navigation inside the ILC is soft, it also uses **2-tiered routing**.

Reviewer 3: ```suggestion
In ILC, you can use one HTML template for all of your applications. With this approach, page load occurs only once (during the first time load), after that all navigation occurs via CSR (client-side rendering).
```

> In addition to the fact that all navigation inside the ILC is soft, it also uses **2-tiered routing**.

Two parts of the sentence are not cohesive. From the first part, it is impossible to guess what's the deal of using 2-tiered routing that you've emphasized. Please complete the sentence by adding a short description of what's the benefit of using 2-tiered routing.

>  Unlike the **flat routing** approach, where we would have to specify in the ILC the full route to every page of all our applications, with a **2-tiered routing** approach, we just need to specify in the ILC route to the application. All navigation within the application can be implemented by each developer team using the native tools of their application (like react-router, vue-router etc.).

Reviewer 2: Unlike the **flat routing** approach, where we would have to specify in the ILC the full route to every page of all our applications, with a **2-tiered routing** approach, we just need to specify in the ILC route to the application.

This statement doesn't make any sense to me....

Reviewer 2: pls rephrase

>  Unlike the **flat routing** approach, where we would have to specify in ILC full URL to every page of all our applications, with the **2-tiered routing** approach, it is enough for us to specify in the ILC, base URL to the application. All navigation within the application can be implemented by each developer team using the native tools of their application (like react-router, vue-router etc.).

Reviewer 3: ```suggestion
Unlike the **flat routing** approach, where, in ILC, you need to specify a full URL to every page of all your applications, with the **2-tiered routing** approach, in ILC, you need to specify only a base URL to the application. All navigation within the application can be implemented using the native tools (for example, react-router, or vue-router).
```

> To make it clearer, here is a small example:

Reviewer 3: ```suggestion
```

 !["2-tiered routing" approach](../assets/2_tiered_routing.png)

>  In this example the user opened a page at `/news/latest` URL. ILC looks at the first part of the URL(`/news/`) to determine what application is it for. It correlates to the `/news/*` route configured in ILC, this route contains information about apps that should be loaded on the page and props they need to receive. When we load and mount the application to its container DOM node – we also pass `basePath` property which should be used by application's router during its work. The application's router processes the complete URL to find the correct page inside its single-page application.

Reviewer 3: ```suggestion
In the example above, the user opened the `/news/latest` page URL. ILC checks the first part of the URL (`/news/`) to determine the assosiated application. It correlates to the `/news/*` route configured in ILC. This route contains information about applications that should be loaded on the page and props these applications need to receive. When the application is loaded and mounted to its container DOM node, it also receives `basePath` property that should be used by application's router. The application's router processes the complete URL to find the correct page inside its SPA.
```

>  Within the application, we can use our native tools (for example `<Link>` in `React router`) to navigate between pages within the application, and `global link` - it's just <a>, tags to navigate between applications.

Reviewer 3: ```suggestion
You can use native tools (for example `<Link>` in `React router`) to navigate between pages within the application, and `global link` - a link (`<a>` tag) to navigate between applications.
```

>  As mentioned before, transition between applications in ILC occurs thanks to the `<a>` tags. To do this, ILC keeps track of all `<a>` tags on the page and handles clicking on them, provided that:

Reviewer 3: ```suggestion
In ILC, as mentioned before, the transition between applications occurs via the `<a>` tags. To do this, ILC keeps track of all `<a>` tags on the page and handles clicks on them, provided that:
```

>  1. tag contains non-empty `href`.
>  2. `event.PreventDefault` not equal `false`
>  3. `target` not equal `_self`
>  4. This is not a special url (`mailto`, `tel`, etc.)

Reviewer 3: ```suggestion
1. Tag contains a non-empty `href`.
1. `event.PreventDefault` does not equal `false`.
1. `target`does not equal `_self`.
1. This is not a special url (`mailto`, `tel`, etc).
```

>  Otherwise, the ILC does not take any part in processing the click on the link.

Reviewer 3: ```suggestion
If one of the above points is not met, ILC ignores the processing of the clicks on the link.
```

> Now let's recap:

Reviewer 3: ```suggestion
Conclusion
```

>  ILC acts as an wrapper for other applications, making all our transitions soft. In addition, we use two-level routing, so that each team can configure routing inside their application as they like, in ILC, you only need to specify the path to the application.

Reviewer 3: ```suggestion
ILC acts as a wrapper for other applications making all the transitions soft. Furthermore, it uses two-level routing so that teams can configure routing inside their application as they need to, whereas, in ILC, you only need to specify the path to the application.
```

The following describes how this is handled in the ILC, but first let's clear up the terminology:

>  - Hard transitions - A transition handled by ILC, with such a transition, the application on the page changes to another one.

Reviewer: "A transition handled by ILC"
"A transition handled by ILC routing"

>  - Soft transitions - A transition that is handled inside the application, with such a transition, only the content inside the application changes, but not the application itself.

Reviewer: To have consistent definitions, instead of "A transition that is handled inside the application" I would use "A transition handled by own routing of some fragment at the page.

And ", but not the application itself" confuses me, so let's remove it.
